plotter/image.go

```diff
@@ -0,0 +1,135 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"image"
+	"image/color"
+	"math"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// Image implements the Plotter interface.
+type Image struct {
+	img            image.Image
+	cols           int
+	rows           int
+	xmin, xmax, dx float64
+	ymin, ymax, dy float64
+}
+
+// NewImage creates a new image plotter.
+func NewImage(img image.Image, xmin, ymin, xmax, ymax float64) *Image {
+	bounds := img.Bounds()
+	cols := bounds.Dx()
+	rows := bounds.Dy()
+	dx := math.Abs(xmax-xmin) / float64(cols)
+	dy := math.Abs(ymax-ymin) / float64(rows)
+	return &Image{
+		img:  img,
+		cols: bounds.Dx(),
+		rows: bounds.Dy(),
+		xmin: xmin,
+		xmax: xmax,
+		dx:   dx,
+		ymin: ymin,
+		ymax: ymax,
+		dy:   dy,
+	}
+}
+
+// Plot implements the Plot method of the plot.Plotter interface.
+func (img *Image) Plot(c draw.Canvas, p *plot.Plot) {
+	trX, trY := p.Transforms(&c)
+	xmin := trX(img.xmin)
+	ymin := trY(img.ymin)
+	xmax := trX(img.xmax)
+	ymax := trY(img.ymax)
+	rect := vg.Rectangle{
+		Min: vg.Point{xmin, ymin},
+		Max: vg.Point{xmax, ymax},
+	}
+	c.DrawImage(rect, img.img)
+}
+
+// DataRange implements the DataRange method
+// of the plot.DataRanger interface.
+func (img *Image) DataRange() (xmin, xmax, ymin, ymax float64) {
+	xmax = img.xmax
+	xmin = img.xmin
+	ymax = img.ymax
+	ymin = img.ymin
+	return xmin, xmax, ymin, ymax
+}
+
+// GlyphBoxes implements the GlyphBoxes method
+// of the plot.GlyphBoxer interface.
+func (img *Image) GlyphBoxes(plt *plot.Plot) []plot.GlyphBox {
+	xmax := img.x(img.cols - 1)
+	xmin := img.x(0)
+	ymax := img.y(img.rows - 1)
+	ymin := img.y(0)
+
+	xnorm := plt.X.Norm
+	ynorm := plt.Y.Norm
+	b := []plot.GlyphBox{
+		{
+			X: xnorm(xmin),
+			Y: ynorm(ymin),
+			Rectangle: vg.Rectangle{
+				Min: vg.Point{X: -1, Y: -1},
+				Max: vg.Point{X: +1, Y: +1},
+			},
+		},
+		{
+			X: xnorm(xmin),
+			Y: ynorm(ymax),
+			Rectangle: vg.Rectangle{
+				Min: vg.Point{X: -1, Y: -1},
+				Max: vg.Point{X: +1, Y: +1},
+			},
+		},
+		{
+			X: xnorm(xmax),
+			Y: ynorm(ymin),
+			Rectangle: vg.Rectangle{
+				Min: vg.Point{X: -1, Y: -1},
+				Max: vg.Point{X: +1, Y: +1},
+			},
+		},
+		{
+			X: xnorm(xmax),
+			Y: ynorm(ymax),
+			Rectangle: vg.Rectangle{
+				Min: vg.Point{X: -1, Y: -1},
+				Max: vg.Point{X: +1, Y: +1},
+			},
+		},
+	}
+	return b
+}
+func (img *Image) x(c int) float64 {
+	if c >= img.cols || c < 0 {
+		panic("plotter/image: illegal range")
+	}
+	return img.xmin + float64(c)*img.dx
+}
+
+func (img *Image) y(r int) float64 {
+	if r >= img.rows || r < 0 {
+		panic("plotter/image: illegal range")
+	}
+	return img.ymin + float64(r)*img.dy
+}
+
+func (img *Image) z(c, r int) color.Color {
+	// image.Image.At(0,0) returns the upper-left pixel
+	// gonum/plot expects 0,0 to be at the bottom-left.
+	v := img.img.At(c, img.rows-1-r)
```

Comment:
actually, `plotter.Image.z` is not needed anymore as we don't fallback on the poor man's raster-in-vg approach.
removed.